### PR TITLE
Added support for custom tags

### PR DESCRIPTION
**Pull request by RandomlyKnighted**
Per #1 this PR adds support for custom tags. I'm sure there might be a more efficient method to do it but this is the method I thought of when I reviewed our emails.

Also in the last commit, I reset the scripts tags to use the full wysiwyg editor but left the development buttons that I added for testing purposes.

### CODE REVIEW

**Comment by chrismichaels84** [comment]
When you are finished, can you remove the console logging.

**Comment by RandomlyKnighted** [comment]
@chrismichaels84 console logging has been removed

**Comment by RandomlyKnighted** [comment]
@chrismichaels84 I believe I've got this finished now. I removed the console logging, removed my comments since I'm assuming you'll be adding more detailed comments (per our previous discussion), and incremented the version number to 2.0.2.

**Comment by chrismichaels84** [comment]
Well, I've pulled this down and played with it in a couple browsers and found some issues we hadn't considered.
1. It doesn't seem to work in IE 11, though I haven't done any investigation as to why
1. There is no way at present to **remove** the custom tags. So, when you select some text and wrap it in custom tags and then click the button again to remove it, the editor wraps the text in a second set of the same tags (sad panda).

I think a lot of this stems from the basic problem with execCommand() as detailed here: http://www.barneyparker.com/the-problem-with-execcommand/

Basically (and I hadn't realized this when I forked the original repo) different browsers implement the different commands in different ways. So firefox wraps bold text in <bold> while IE wraps the text in <strong>. If you begin editing in firefox, save and then try to edit next time in IE, strange things will happen.

Not sure the best way to solve either of these problems.

**Comment by RandomlyKnighted** [comment]
1. Odd, alright I'll see if I can figure out what the issue is  IE11.
2. After a quick Google search, [this seems promising](http://stackoverflow.com/questions/14028773/javascript-execcommandremoveformat-doesnt-strip-h2-tag).
3. I remember reading about that, but I didn't think about saving then editing later. I'm not exactly sure how we might handle something like that. Do you have any suggestions?

**Comment by RandomlyKnighted** [comment]
1. It appears that in IE11 support for ```document.getSelection()``` was dropped and the insertHTML feature of execCommand has still not been implemented, which would explain why it doesn't work in IE11.
2. I haven't got this far yet. I'm still working to make what we already have work with IE11..though my current thinking is that once we get it working in IE11 we could test to see if the selection is already wrapped in tags. If so then we can use the unwrap function to remove them rather than wrapping them again.
3. I'm still working out the details on this, but everything I've thought of so far would be far from a simple solution to this particular problem.

**Comment by chrismichaels84** [comment]
I have also been playing with this over the weekend, and I agree. These are not easy problems. My first thought was to open source a simple normalizer that would replace the execCommand for commands where the implementation is different. There are libraries out there (ckeditor) that don't use execCommand, but they are terribly complex.

I did realize though (at least on my machine) that most of the browsers will work fine if you begin in one and edit in another. The end result works well, but there are a few bugs in the markup (like extra `<strong>` tags laying around). So it may be possible just to clean things up.

As far as toggling, that seems the best way to go, but it can be complicated when you have a situation like `<b>This is my text</b>` and you want to unbold only `my` so you select it. In that case the library would have to realize it should add an end bold and then a begin bold tag. There is also the problem with nested markup `<b><i>This</i> is <u><i>my</u> terrible</i> test</b>` Something like that may make the library explode.

Truth be told, in my production apps, I would probably opt for something like aloha or ckeditor. I'd still like to be able to provide a lightweight option, but I'm running out of ideas.

**Comment by RandomlyKnighted** [comment]
I started to look at how other's handled the situations and found a couple things of interest. Both come from the same, as you can [see here](https://github.com/adamsanderson/wysiwyg/blob/gh-pages/build/build.js#L82) these developers made a function to normalize the text. Second, I'm sure you [noticed the commands array](https://github.com/adamsanderson/wysiwyg/blob/gh-pages/build/build.js#L233) that shows what to pass to the execCommand. I like this because with the problems we mentioned we could something like:

```
var commands = {
  bold:        wrapWithTag("b"),
  indent:     makeCommand("indent"),
```

where the wrapWithTag command is used when there a command is used that is implemented differently depending on the browser. 

I think the normalizer would be more complicated because you'd have to handle various scenarios, but the other option means we'd have to handle all everything that the browsers do which is something you said you didn't want to do. Either way I think both options would work.

**Comment by chrismichaels84** [comment]
This all looks great. I had never seen this repo and had never encountered `queryCommandValue()` which seems to provide a way to know if selected text is bold, highlighted, whatever. That could allow us to normalize things rather easily (relatively).

**Comment by RandomlyKnighted** [comment]
So what would be your recommendations moving forward?

**Comment by chrismichaels84** [comment]
Do you have any thoughts?

**Comment by chrismichaels84** [comment]
I don't know. I'd like to look through that repo more, but I can't today. Probably canibalize that repo for whatever we can. Once its stable, I'd probably like to extract part of it to another repo and include it. 

In the fewest lines of code possible, what do you think about this:

  * For the execCommand():
```js
var commands = { /* only for custom commands or those with wildly different implementation across the main browsers */
    'bold': wrapWithTags(['<b>', '</b>']), // wrap the selection with tags
    'red': styleAs('color', 'red'), //wrap selection in <span style="color: red"></span>. Should use wrapWithTags()
}

if (commands.has(command) {
    // Use the override command
    var isApplied = x; /* figure out if the current selection is already "bold" or whatever the command is. (maybe with queryCommandValue(); */

    var isFragment = x; /* like selecting only "a little" from <b>Mary had a little lamb</b> which would unbold a little */

   if (isApplied && isFragment) {
      // wrap the selection in the closing tag and then the opening tag. Easy by reversing the array and overwriting the selction
   } else if (isApplied && ! isFragment) {
      // remove the tags completely. Less easy.
   } else {
      // Not applied, so wrap the selection in tags. Also easy, I would think
   }
} else {
   // use the regular document.execCommand();
}
```

With something like this at the core, we should be able to easily use the same api that we worked out and I like a lot (<a data-edit="custom p|code">)

But, I'm not sure how to find `isApplied` and `isFragment`. Or if this is even the best way to go, honestly.

**Comment by chrismichaels84** [comment]
Take a look at https://github.com/timdown/rangy, https://github.com/timdown/rangyinputs, and especially the class-applier module for rangy. Right now, the class applier toggles a `span` tag to apply classes. I peeked through the source code and it shouldn't be difficult to make it possible to toggle any valid element. Rangy inputs is tiny and has a surroundContents() that works wonderfully. It may be possible to cannibalize these libraries to add a toggleSurroundContents() to rangyinputs from class-applier?

I don't want to add a lot of dependencies to this. There are already so many wysiwyg editors, I just wanted this to be a dead simple and light jQuery plugin. Maybe these changes could be put into a small, optional module? You load the optional module just after the library itself and it overrides the libraries execCommand() method.

Thoughts?

**Comment by RandomlyKnighted** [comment]
I haven't had a chance to go through what you gave in a lot of detail yet. However, while working on one of my main projects I've been learning gulp and so I decided to use it to check the source for errors using JSHint. My last commit fixes some the errors I found.

**Comment by RandomlyKnighted** [comment]
@chrismichaels84 unless I'm misinterpreting the documentation it looks like ```document.queryCommandEnabled()``` will allow us to determine if the selected text is currently bold, underlined, etc. This is already in use in the updateToolbar function if you need to see an example of it. If you're still interested in adding a normalizer.

Per your last post, honestly, I'd prefer not to add anymore dependencies than we already have. If we do then we start getting a way from being an (almost - due to the jquery.hotkeys dependency) pure jQuery project. The less we dependencies we have the easier it will be to use, but that's just my 2 cents.

**Comment by chrismichaels84** [comment]
I agree wholeheartedly. I do not want to add anymore dependencies, and have even considered removing jQuery hotkeys and including it into the core. The repo hasn't been updated in a while and its pretty small.

I don't want to add Rangy. It's huge. I am glad to hear that we're on the same page with `document.queryCommandEnabled()`, that is also how I read the documentation. Our own version of surroundContents using queryCommandEnabled() would probably be the best bet.

I'm leaving for the weekend. I can tackle a code-sketch next week, or if you want to take a look at how this would work that would be great. I really appreciate your help on this project.

I have also done some looking into the execCommand normalizer. The problem isn't as bad as I thought, at least on relatively current browsers. You can see my thoughts in #5. I've dropped it for now.

**Comment by RandomlyKnighted** [comment]
Were you able to make a code sketch of this?

**Comment by chrismichaels84** [comment]
To be honest man, I don't think I'm able to dedicate any real time to this
package right now. I have some deadlines, and I just can't spend my on
something I'm not using at the moment. I'm happy to currate and commit PRs
and be engaged in conversation. But, for the moment, I need to work on some
backend projects in python and php and the context switching is making my
brain go numb.

>
